discord/asset.py

```diff
@@ -0,0 +1,144 @@
+
+        .. describe:: bool(x)
+
+            Checks if the Asset has a URL.
+    """
+    __slots__ = ('_state', '_url')
+
+    def __init__(self, state, url=None):
+        self._state = state
+        self._url = url
+
+    @classmethod
+    def _from_avatar(cls, state, user, *, format=None, static_format='webp', size=1024):
+        if not utils.valid_icon_size(size):
+            raise InvalidArgument("size must be a power of 2 between 16 and 1024")
+        if format is not None and format not in VALID_AVATAR_FORMATS:
+            raise InvalidArgument("format must be None or one of {}".format(VALID_AVATAR_FORMATS))
+        if format == "gif" and not user.is_avatar_animated():
+            raise InvalidArgument("non animated avatars do not support gif format")
+        if static_format not in VALID_STATIC_FORMATS:
+            raise InvalidArgument("static_format must be one of {}".format(VALID_STATIC_FORMATS))
+
+        if user.avatar is None:
+            return user.default_avatar_url
+
+        if format is None:
+            format = 'gif' if user.is_avatar_animated() else static_format
+
+        return cls(state, 'https://cdn.discordapp.com/avatars/{0.id}/{0.avatar}.{1}?size={2}'.format(user, format, size))
+
+    @classmethod
+    def _from_icon(cls, state, object, path):
+        if object.icon is None:
+            return cls(state)
+
+        url = 'https://cdn.discordapp.com/{0}-icons/{1.id}/{1.icon}.jpg'.format(path, object)
+        return cls(state, url)
+
+    @classmethod
+    def _from_guild_image(cls, state, id, hash, endpoint, *, format='webp', size=1024):
+        if not utils.valid_icon_size(size):
+            raise InvalidArgument("size must be a power of 2 between 16 and 4096")
+        if format not in VALID_STATIC_FORMATS:
+            raise InvalidArgument("format must be one of {}".format(VALID_STATIC_FORMATS))
+
+        if hash is None:
+            return Asset(state)
+
+        return cls(state, endpoint.format(id, hash, format, size))
+
+    def __str__(self):
+        return self._url
+
+    def __len__(self):
+        return len(self._url)
+
+    def __bool__(self):
+        return self._url is not None
+
+    def __repr__(self):
+        return '<Asset url={0._url!r}>'.format(self)
+
+    async def save(self, fp, *, seek_begin=True):
+        """|coro|
+
+        Saves this asset into a file-like object.
+
+        Parameters
+        -----------
+        fp: Union[BinaryIO, :class:`os.PathLike`]
+            Same as in :meth:`Attachment.save`.
+        seek_begin: :class:`bool`
+            Same as in :meth:`Attachment.save`.
+
+        Raises
+        --------
+        HTTPException
+            Saving the asset failed.
+        NotFound
+            The asset was deleted.
+
+        Returns
+        --------
+        :class:`int`
+            The number of bytes written.
+        """
+        if not self._url:
+            raise DiscordException('a URL was not provided')
```

Comment:
This exception is undocumented.

```suggestion
            raise DiscordException('Invalid asset (no URL provided)')
```